pageserver/src/disk_usage_eviction_task.rs

```diff
@@ -591,12 +709,33 @@ async fn collect_eviction_candidates(
         tenant_candidates
             .sort_unstable_by_key(|(_, layer_info)| std::cmp::Reverse(layer_info.last_activity_ts));
         let mut cumsum: i128 = 0;
-        for (timeline, layer_info) in tenant_candidates.into_iter() {
+
+        // keeping the -1 or not decides if every tenant should lose their least recently accessed
+        // layer OR if this should happen in the order of having highest layer count:
+        //
+        // relative_age vs. tenant layer count without - 1:
+        // - 0.1..=1.0 (10 layers)
+        // - 0.01..=1.0 (100 layers)
+        // - 0.001..=1.0 (1000 layers)
+        let fudge = if eviction_order.highest_layer_count_loses_first() {
+            0
+        } else {
+            1
+        };
+
+        let total = tenant_candidates.len() - fudge;
+        let divider = total as f32;
+
+        for (i, (timeline, layer_info)) in tenant_candidates.into_iter().enumerate() {
             let file_size = layer_info.file_size();
             let candidate = EvictionCandidate {
                 timeline,
                 last_activity_ts: layer_info.last_activity_ts,
                 layer: layer_info.layer,
+                relative_last_activity: finite_f32::FiniteF32::try_from_normalized(
+                    (total - i) as f32 / divider,
```

Comment:
divider can become 0 if `tenant_candidates.len() == 1` and `fudge == 1`. then we panic here.

Reply:
also, this seems to be achievable with `std::cmp::Reverse`, or using integers and not dividing by `divider` (the total isn't even needed we can just use `-(i as isize)`). if avoidable, we should not be using floats as they have many issues.

Reply:
This sounds like a great idea, also thanks for finding the panic.

Reply:
Could you elaborate on the idea of using just the `-(i as isize)`, how would it achieve normalizing so that what is described in the example for this function happens? I think we must divide with something.

Reply:
https://github.com/neondatabase/neon/pull/6136#discussion_r1432840994